Crawl-delay in robots.txt should not shrink delay configured by fetcher.server.delay

sebastian-nagel opened the pull request:
to guarantee that the crawler always behaves polite and inconspicuous, see commoncrawl/news-crawl#24.

sebastian-nagel commented:
PR allows avoids repeated messages `Crawl delay for queue: ...  is set to 1000 as per robots.txt.`

jnioche commented on core/src/main/java/com/digitalpebble/stormcrawler/bolt/FetcherBolt.java:
```diff
@@ -402,6 +404,8 @@ public FetcherThread(Config conf, int num) {
 
             this.maxCrawlDelay = ConfUtils.getInt(conf,
                     "fetcher.max.crawl.delay", 30) * 1000;
+            this.crawlDelay = (long) (ConfUtils.getFloat(conf,
```
probably don't need to read this but instead use fiq.crawlDelay as you do in line 534

sebastian-nagel commented:
Agreed. What about a property `fetcher.min.crawl.delay`? The crawl delay will then be between this value and that of `fetcher.max.crawl.delay`. There is already a `fetcher.server.min.delay` but it's used as delay between "fetches without fetch" (eg. robots.txt denied).

jnioche commented:
>  There is already a fetcher.server.min.delay but it's used as delay between "fetches without fetch" (eg. robots.txt denied).

No, it is used in normal fetches (`asap==false`). Robots denied have the opposite value of `asap=true`. 
It is used when multithreading on a single queue to define an alternative value to the normal crawl delay. This is inherited from Nutch and I agree that it is confusing. The name of the param doesn't help either (server?)

Let me have a think about it. Do we need a special case for multithreading or would the `fetcher.min.crawl.delay` be used no matter what? It could always be set to 0 or a low value for users who want an aggressive setting. If so should we declare it in the same way and place as the max value?  

Your thoughts and suggestions are welcome as usual

sebastian-nagel commented:
Ok, sorry, you're right.
> Do we need a special case for multithreading
It could make sense if there is a custom per-queue number of threads (`fetcher.maxThreads.queueId`): then  `fetcher.server.min.delay` is for queues which you're explicitly want aggressive crawling while the other is to guarantee a minimum delay for the polite queues.

sebastian-nagel commented:
Added a property `fetcher.min.crawl.delay`. Removed change from SimpleFetcherBolt which simply ignores overlong crawl delays, so it should be ok with shorter delays.

jnioche commented:
If users are concerned with being too aggressive even if permitted by robots, would they set a min value different from the default delay they specify? What about having a boolean parameter instead indicating that if a delay from robots is found and is lower than the default, we should use the latter anyway? It would be set to false by default.

As for SimpleFetcherBolt, the logic is different from FetcherBolt when it comes to the max values (but that won't necessarily be always the case) and we could have the same logic as FetcherBolt when it comes to min values. In fact, we could have a new config indicating whether we should skip a URL if the max value set by a server is above our maxValue (as currently done in FetcherBolt) or enforce the max value we set (as in SimpleFetcherBolt)

sebastian-nagel commented:
> would they set a min value different from the default delay they specify?

Rarely. And it makes even more sense if there is also a boolean whether to enforce the max. configured delay.

How to name the new boolean properties? fetcher.server.delay.enforce resp. fetcher.max.crawl.delay.enforce?

jnioche commented:
or just fetcher.server.delay.force and fetcher.max.crawl.delay.force ?

it will be important to document the behaviour of these configs in the
default config, code and WIKI as their meaning is not necessarily very
obvious nor do they have the same semantic between them

Thanks!


On 21 March 2018 at 13:37, Sebastian Nagel <notifications@github.com> wrote:

> would they set a min value different from the default delay they specify?
>
> Rarely. And it makes even more sense if there is also a boolean whether to
> enforce the max. configured delay.
>
> How to name the new boolean properties? fetcher.server.delay.enforce resp.
> fetcher.max.crawl.delay.enforce?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/DigitalPebble/storm-crawler/pull/549#issuecomment-374938865>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANUzwGN600lt8xNsY6Qavg_bzcrOowSks5tglergaJpZM4St8wT>
> .
>



-- 



* Open Source Solutions for Text Engineering   http://www.digitalpebble.com
<http://www.digitalpebble.com>*
*http://digitalpebble.blogspot.com <http://digitalpebble.blogspot.com/>*
@digitalpebble <https://twitter.com/digitalpebble>

sebastian-nagel commented:
ok, I'll update the PR tomorrow. Thanks!

sebastian-nagel commented:
Done: properties `fetcher.server.delay.force` and `fetcher.max.crawl.delay.force` are added in ad15f72, [description for wiki](https://github.com/sebastian-nagel/storm-crawler/wiki/Configuration/_compare/0e378bfa80af8e1a7007e80c8175dfd02b3caf27...f221174f16a756791117836113a89611fb9458cc).

jnioche commented:
Fab, thanks @sebastian-nagel
